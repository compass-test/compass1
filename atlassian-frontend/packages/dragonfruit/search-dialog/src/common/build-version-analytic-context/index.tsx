export { BuildVersionAnalyticContext } from './build-version-analytic-context';
