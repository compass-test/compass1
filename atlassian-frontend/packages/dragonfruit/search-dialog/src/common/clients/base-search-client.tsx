import {
  RequestServiceOptions,
  ServiceConfig,
  utils,
} from '@atlaskit/util-service-support';
import {
  SearchContext,
  ABTest,
  ScopedAggregatorResponse,
  ModelParam,
  Site,
} from './common-types';
import { Products } from '../product-context';
import { timed } from '../../common/clients/timing';
import { limitToMaxSiteUsage } from './multi-site-utils';

export interface SearchParams<SearchScope, SupportedFilters> {
  query: string;
  context: SearchContext;
  scopes: SearchScope[];
  modelParams: ModelParam[];
  resultLimit?: number;
  filters?: SupportedFilters[];
  experience: string;
  /**
   * A list of cloud ids to search. If this is empty we will use the default `cloudId` that is used to construct this client to perform the search.
   */
  sites?: string[];
}

// Type used to help unpack the types that make up an array
type Unpack<A> = A extends Array<infer E> ? E : A;
// Type used to lookup the response based on the scope
type ResponseFromScope<S, T> = S extends ScopedAggregatorResponse<infer E>
  ? T extends E
    ? S
    : never
  : never;

interface MultiscopeSearchResponse<T> {
  scopes: T[];
}

export interface SearchResponse<
  Scope,
  Response extends ScopedAggregatorResponse<Scope>,
  SearchScope
> {
  rawData: MultiscopeSearchResponse<
    ResponseFromScope<Response, Unpack<SearchScope>>
  >;
  retrieveScope<SingleScope extends Unpack<SearchScope>>(
    scope: SingleScope,
  ): ResponseFromScope<Response, SingleScope> | null;
}

export interface SearchResponseWithTiming<
  S,
  R extends ScopedAggregatorResponse<S>,
  SR
> {
  response: SearchResponse<S, R, SR>;
  requestDurationMs: number;
}

interface ScopedAbTestResponse<SupportedScopes> {
  id: SupportedScopes;
  error?: string;
  abTest: ABTest;
}

interface MultiscopeAbTestResponse<SupportedScopes> {
  scopes: ScopedAbTestResponse<SupportedScopes>[];
}

export interface ProductsPermissionsResponse<SupportedScopes> {
  products: { [key in keyof typeof Products]: SupportedScopes[] };
}

export interface AggregatorConfig {
  cloudId: string;
  siteMasterList: Site[];
  url: string;
}

export type ServerProduct = 'jira' | 'confluence';

export class AggregatorClient<
  Resp extends ScopedAggregatorResponse<Scope>,
  SupportedFilters,
  Scope = Resp['id']
> {
  private serviceConfig: ServiceConfig;

  private cloudId: string;

  private siteMasterList: Site[];

  // Static URLs
  private static readonly EXPERIMENT_URL_PATH = 'experiment/v1';

  private static readonly QUICKSEARCH_API_URL = 'quicksearch/v1';

  private static readonly SCOPES_API_URL = 'scopes/v1';

  constructor({ url, cloudId, siteMasterList }: AggregatorConfig) {
    this.serviceConfig = { url: url };
    this.cloudId = cloudId;
    this.siteMasterList = siteMasterList;
  }

  public async getAbTestData(
    scope: Scope,
    experience: string,
  ): Promise<ABTest | null> {
    const body = {
      cloudId: this.cloudId,
      scopes: [scope],
      experience,
    };

    const response = await this.makeRequest<MultiscopeAbTestResponse<Scope>>(
      AggregatorClient.EXPERIMENT_URL_PATH,
      body,
    );

    const scopeWithAbTest:
      | ScopedAbTestResponse<Scope>
      | undefined = response.scopes.find((s) => s.id === scope);

    const abTestPromise = scopeWithAbTest
      ? Promise.resolve(scopeWithAbTest.abTest)
      : Promise.resolve(null);

    return abTestPromise;
  }

  public async getProductPermissions(
    products: Products[],
    experience: string,
  ): Promise<Products[]> {
    if (products.length === 0) {
      return Promise.resolve(products);
    }

    try {
      const response = await this.makeRequest<
        ProductsPermissionsResponse<Scope>
      >(AggregatorClient.SCOPES_API_URL, {
        cloudId: this.cloudId,
        productIds: products.map((product) => product.toLowerCase()),
        experience,
      });

      const allowedProductKeys = Object.keys(response.products) as Array<
        keyof typeof Products
      >;

      return Promise.resolve(
        allowedProductKeys
          .filter((product) => response.products[product].length > 0)
          .map((product) => product.toUpperCase() as Products),
      );
    } catch (e) {
      return Promise.resolve([]);
    }
  }

  public async search<SearchScope extends Scope>({
    query,
    context,
    scopes,
    modelParams,
    resultLimit,
    filters = [],
    experience,
    sites = [],
  }: SearchParams<SearchScope, SupportedFilters>): Promise<
    SearchResponseWithTiming<Scope, Resp, SearchScope>
  > {
    const cloudId = this.cloudId;
    const cloudIdsToFilterOn =
      sites.length > 0
        ? sites
        : limitToMaxSiteUsage(this.siteMasterList).map((site) => site.cloudId);
    const body = {
      query: query,
      cloudId,
      limit: resultLimit,
      scopes,
      filters: filters,
      searchSession: context,
      ...(modelParams.length > 0 ? { modelParams } : {}),
      experience,
      cloudIds: cloudIdsToFilterOn,
    };

    const { result: response, durationMs: requestDurationMs } = await timed(
      this.makeRequest<
        MultiscopeSearchResponse<ResponseFromScope<Resp, Unpack<SearchScope>>>
      >(AggregatorClient.QUICKSEARCH_API_URL, body),
    );

    return {
      response: {
        rawData: response,
        retrieveScope: (scope) =>
          response.scopes.find((s) => s.id === scope) || null,
      },
      requestDurationMs,
    };
  }

  private async makeRequest<T>(path: string, body: object): Promise<T> {
    const options: RequestServiceOptions = {
      path,
      requestInit: {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      },
    };

    const results = await utils.requestService<T>(this.serviceConfig, options);

    return results;
  }
}
