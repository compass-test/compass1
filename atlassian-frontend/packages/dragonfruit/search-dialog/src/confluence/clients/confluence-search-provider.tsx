import React, { FunctionComponent } from 'react';
import {
  ConfluenceSearchClient,
  ConfluenceSearchClientConfig,
} from './confluence-search-client';
import memo from 'memoize-one';
import deepEqual from 'deep-equal';
import {
  CollaborationGraphConfig,
  CollaborationGraphClient,
  Site,
  extractSiteUrl,
} from '../../common/clients';
import { ConfItemResults } from './response-types';

export interface SearchClientContext {
  searchClient: ConfluenceSearchClient;
  collabGraphClient: CollaborationGraphClient;
  siteUrl: string;
  cloudId: string;
  sites: Array<Site>;
}

export interface SearchClientConfig {
  /**
   * The url of the aggregator to call. This should be /gateway/api/xpsearch-aggregator unless there's a good reason to override this (e.g. in tests)
   */
  aggregatorUrl: string;

  /**
   * The base url of the instance.
   */
  baseUrl: string;

  /**
   * This represents the instance to search against when `sites` is not provided. When `sites` is provided this will be used to provide contextual information for the search.
   */
  cloudId: string;

  /**
   * Overrides the default url to the collaboration graph. Typically this should only be used for tests.
   */
  collaborationGraphUrl?: string;

  /**
   * Whether the user is anonymous. Search results may differ for anonymous user.
   */
  isUserAnonymous: boolean;

  /**
   * An override to provide a custom getter for retrieving recent confluence pages, blogs and attachments.
   */
  recentItemsSupplier?: () => Promise<ConfItemResults>;

  /**
   * Whether to use collaboration graph results instead of Confluence's recents endpoint to determine recent items.
   *
   * This should be set to true when used outside of a Jira or Confluence context.
   */
  useCollaborationGraphForRecents?: boolean;

  /**
   * The site url of the instance without a trailing '/'
   */
  siteUrl?: string;

  /**
   * A list of sites to search against. If this is not provided the `cloudId` will be used to determine what instance to search against.
   */
  sites?: Site[];
}

const defaultCollaborationGraphConfig = {
  collaborationGraphUrl: 'gateway/api/collaboration',
  useCollaborationGraphForRecents: false,
};

const getSearchClientWithConfig = memo(
  (
    config: ConfluenceSearchClientConfig,
    collabGraphClient: CollaborationGraphClient,
  ) => new ConfluenceSearchClient(config, collabGraphClient),
  deepEqual,
);

const getCollabGraphClientWithConfig = memo(
  (config: CollaborationGraphConfig) =>
    new CollaborationGraphClient({
      collaborationGraphUrl: config.collaborationGraphUrl,
      cloudId: config.cloudId,
      isMultiSite: config.isMultiSite,
      sites: config.sites,
    }),
  deepEqual,
);

export const ConfluenceSearchClientsContext = React.createContext<
  Partial<SearchClientContext>
>({});

const MISSING_PROVIDER_ERROR = new Error(
  `Could not find all required Clients, make sure ConfluenceClientsProvider is created further up the tree`,
);

export interface ConfluenceClientsSupplierProps {
  children: (clients: SearchClientContext) => JSX.Element;
}

export const withClients = <T extends Partial<SearchClientContext>>(
  Component: React.ComponentType<T>,
) => {
  type PropsExcludedClients = Omit<T, 'searchClient' | 'collabGraphClient'>;

  return (props: PropsExcludedClients) => (
    <ConfluenceSearchClientsContext.Consumer>
      {({ searchClient, cloudId, siteUrl = '', sites }) => {
        if (!searchClient) {
          throw MISSING_PROVIDER_ERROR;
        }

        const { ...rest } = props;

        return (
          <Component
            {...(rest as any)}
            searchClient={searchClient}
            cloudId={cloudId}
            siteUrl={siteUrl}
            sites={sites}
          />
        );
      }}
    </ConfluenceSearchClientsContext.Consumer>
  );
};

export interface ConfluenceClientsProviderProps {
  config: SearchClientConfig;
}

export const ConfluenceClientsProvider: FunctionComponent<ConfluenceClientsProviderProps> = ({
  children,
  config,
}) => {
  const {
    cloudId,
    collaborationGraphUrl,
    aggregatorUrl,
    isUserAnonymous,
    useCollaborationGraphForRecents,
    sites = [],
  } = {
    ...defaultCollaborationGraphConfig,
    ...config,
  };

  const siteUrl = extractSiteUrl(config);

  const collabGraphClient = getCollabGraphClientWithConfig({
    collaborationGraphUrl: collaborationGraphUrl,
    cloudId: cloudId,
    isMultiSite: sites.length > 1,
    sites,
  });

  return (
    <ConfluenceSearchClientsContext.Provider
      value={{
        searchClient: getSearchClientWithConfig(
          {
            isUserAnonymous: isUserAnonymous,
            url: aggregatorUrl,
            cloudId: cloudId,
            isCollaborationGraphEnabled: useCollaborationGraphForRecents,
            siteMasterList: sites,
          },
          collabGraphClient,
        ),
        collabGraphClient: collabGraphClient,
        siteUrl,
        cloudId,
        sites,
      }}
    >
      {children}
    </ConfluenceSearchClientsContext.Provider>
  );
};

export const useClients = () => {
  const context = React.useContext(ConfluenceSearchClientsContext);

  if (!context.searchClient) {
    throw MISSING_PROVIDER_ERROR;
  }

  return context as SearchClientContext;
};
