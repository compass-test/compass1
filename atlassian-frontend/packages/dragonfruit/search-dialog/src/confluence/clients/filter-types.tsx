/**
 * All filters must implement this interface.
 */
export interface FilterInterface {
  '@type': string;
}

export interface SpaceFiltersInterface {
  '@type': 'spaces';
  spaceKeys: string[];
}

export interface UserFiltersInterface {
  '@type': 'contributors';
  accountIds: string[];
}

export type AllFilters = SpaceFiltersInterface | UserFiltersInterface;
