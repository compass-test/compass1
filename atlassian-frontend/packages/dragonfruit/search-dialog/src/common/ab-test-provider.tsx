import React, {
  useContext,
  FunctionComponent,
  useState,
  useEffect,
} from 'react';
import { ABTest, SharedClient } from './clients';
import { SearchDialogAnalyticsContext, useAnalytics } from './analytics';
import { onExperimentLoaded } from './analytics/events';
import { ProductSearchInputSkeleton } from './product-search-input-skeleton';

export const DEFAULT_AB_TEST: ABTest = {
  abTestId: 'default',
  controlId: 'default',
  experimentId: 'nav-v3',
};

const MISSING_PROVIDER_ERROR = new Error(
  'Experimentation data has not been initialised, please ensure that there is an ABTestContext.Provider available',
);
interface Props {
  searchClient: SharedClient;
  isMultiProduct?: boolean;
}

interface State {
  abTest: ABTest | null;
}

export const ABTestContext = React.createContext<State>({
  abTest: null,
});

export const ABTestProvider: FunctionComponent<Props> = ({
  searchClient,
  isMultiProduct,
  children,
}) => {
  const [abTest, setAbTest] = useState<ABTest | null>(null);
  const { fireAnalyticsEvent } = useAnalytics();

  useEffect(() => {
    const now = Date.now();

    searchClient
      .getAbTestData()
      .then((abTestResponse) => {
        const duration = Date.now() - now;
        const abTestWithDefault = abTestResponse || DEFAULT_AB_TEST;
        setAbTest(abTestWithDefault);
        fireAnalyticsEvent(
          onExperimentLoaded(abTestWithDefault, duration, !!isMultiProduct),
        );
      })
      .catch((e) => {
        setAbTest(DEFAULT_AB_TEST);
      });
  }, [searchClient, fireAnalyticsEvent, isMultiProduct]);

  if (!abTest) {
    return <ProductSearchInputSkeleton />;
  }

  return (
    <ABTestContext.Provider value={{ abTest }}>
      <SearchDialogAnalyticsContext
        analyticContext={{ abTest }}
        nonPrivacySafeAnalyticContext={{}}
      >
        {children}
      </SearchDialogAnalyticsContext>
    </ABTestContext.Provider>
  );
};

export const useABTest = () => {
  const { abTest } = useContext(ABTestContext);

  if (!abTest) {
    throw MISSING_PROVIDER_ERROR;
  }

  return abTest;
};
