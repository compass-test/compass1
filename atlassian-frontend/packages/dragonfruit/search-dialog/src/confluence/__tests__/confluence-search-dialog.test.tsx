import React from 'react';

import {
  WrappedConfluenceSearchDialog,
  ConfluenceSearchDialog,
  State,
} from '../confluence-search-dialog';
import { shallow, ShallowWrapper } from 'enzyme';

import {
  createPageBlogAttachmentResults,
  createPeopleResults,
  createSpaceResponse,
} from '../../__tests__/__fixtures__/mock-search-results';
import { SearchError } from '../../common/search-error';
import { Scope } from '../clients';
import CancellablePromise from '../../utils/cancellable-promise';
import { createSiteFilters } from '../../__tests__/__fixtures__/mock-filters';
import { CompassComponentType } from '@atlassian/dragonfruit-graphql';

import {
  fireUIAnalytics,
  fireTrackAnalytics,
} from '@atlassian/analytics-bridge';
import { useSearchComponentsLazyQuery } from '@atlassian/dragonfruit-graphql';
import { useSearchUsers } from '@atlassian/dragonfruit-teams';
import { useSearchTeams } from '@atlassian/dragonfruit-rest';

jest.unmock('lodash/debounce');

jest.mock('../../utils/forward-ref', () => ({
  withForwardRef: (arg: any) => arg,
}));

jest.mock('@atlassian/search-dialog', () => {
  const React = require('react');
  return {
    SearchAnchor: () => 'div',
    SearchDialog: () => 'div',
    SearchInput: () => 'div',
    SearchFooter: () => 'div',
    KeyboardHighlightProvider: () => 'div',
    ThemeProvider: React.createContext({}).Provider,
    Link: () => 'a',
  };
});

jest.mock('@atlaskit/width-detector', () => ({
  WidthObserver: () => null,
}));

jest.mock('@atlaskit/analytics-next', () =>
  Object.assign({}, jest.requireActual('@atlaskit/analytics-next'), {
    useAnalyticsEvents: () => ({
      createAnalyticsEvent: jest.fn().mockImplementation((data) => data),
    }),
  }),
);

jest.mock('@atlassian/analytics-bridge', () =>
  Object.assign({}, jest.requireActual('@atlassian/analytics-bridge'), {
    fireTrackAnalytics: jest.fn(),
    fireUIAnalytics: jest.fn(),
  }),
);

jest.mock('@atlassian/dragonfruit-teams', () => ({
  useSearchUsers: jest.fn().mockReturnValue({
    usersLoading: false,
    users: [],
    fetchData: jest.fn(),
  }),
}));

jest.mock('@atlassian/dragonfruit-rest', () => ({
  useGetTeams: jest.fn().mockReturnValue({ teamNames: [] }),
  useSearchTeams: jest.fn().mockReturnValue({
    teams: [],
    loading: false,
    fetchData: jest.fn(),
  }),
}));

jest.mock('@atlassian/compass-search-cache', () =>
  Object.assign({}, jest.requireActual('@atlassian/compass-search-cache'), {
    useCompassRecents: jest.fn().mockReturnValue({
      recentlyViewedComponents: [],
      recentlyViewedTeams: [],
    }),
  }),
);

jest.mock('@atlassian/dragonfruit-tenant-context', () => ({
  useTenantInfo: () => ({
    cloudId: 'mockCloudId',
  }),
  getEnvironmentFromOrigin: () => 'local',
  Environment: () => ({
    PROD: 'prod',
    STAGING: 'staging',
    LOCAL: 'local',
  }),
}));

jest.mock('@atlassian/dragonfruit-graphql', () =>
  Object.assign({}, jest.requireActual('@atlassian/dragonfruit-graphql'), {
    useSearchComponentsLazyQuery: jest.fn().mockReturnValue([jest.fn(), {}]),
  }),
);

describe('<ConfluenceSearchDialog />', () => {
  const getRecentPeople = jest.fn();
  const getRecentItems = jest.fn();
  const getRecentSpaces = jest.fn();

  const searchPeople = jest.fn();
  const searchItems = jest.fn();
  const searchSpaces = jest.fn();
  const fetchMore = jest.fn();

  const mockEvent = { event: 'event' };
  const mockCreateAnalyticsEvent = jest.fn().mockReturnValue(mockEvent);

  const REJECT_PROMISE = Promise.reject('error');

  const recentPeopleResults = createPeopleResults(1);
  const recentItemResults = createPageBlogAttachmentResults(1);
  const recentSpaceResults = createSpaceResponse(1);

  const searchPeopleResults = createPeopleResults(1);
  const searchItemResults = createPageBlogAttachmentResults(1);
  const searchSpaceResults = createSpaceResponse(1);

  const mockSuccessRecent = () => {
    getRecentPeople.mockReturnValue(
      CancellablePromise.from(Promise.resolve(recentPeopleResults)),
    );
    getRecentItems.mockReturnValue(
      CancellablePromise.from(Promise.resolve(recentItemResults)),
    );
    getRecentSpaces.mockReturnValue(
      CancellablePromise.from(Promise.resolve(recentSpaceResults)),
    );
  };
  const mockEmptyRecents = () => {
    getRecentPeople.mockReturnValue(
      CancellablePromise.from(Promise.resolve(createPeopleResults(0))),
    );
    getRecentItems.mockReturnValue(
      CancellablePromise.from(
        Promise.resolve(createPageBlogAttachmentResults(0)),
      ),
    );
    getRecentSpaces.mockReturnValue(
      CancellablePromise.from(Promise.resolve(createSpaceResponse(0))),
    );
  };

  const mockErrorSearch = () => {
    searchPeople.mockReturnValue(CancellablePromise.from(REJECT_PROMISE));
    searchItems.mockReturnValue(CancellablePromise.from(REJECT_PROMISE));
    searchSpaces.mockReturnValue(CancellablePromise.from(REJECT_PROMISE));
  };
  const mockSuccessSearch = () => {
    searchPeople.mockReturnValue(
      CancellablePromise.from(Promise.resolve(searchPeopleResults)),
    );
    searchItems.mockReturnValue(
      CancellablePromise.from(Promise.resolve(searchItemResults)),
    );
    searchSpaces.mockReturnValue(
      CancellablePromise.from(Promise.resolve(searchSpaceResults)),
    );
  };
  const mockEmptySearch = () => {
    searchPeople.mockReturnValue(
      CancellablePromise.from(Promise.resolve(createPeopleResults(0))),
    );
    searchItems.mockReturnValue(
      CancellablePromise.from(
        Promise.resolve(createPageBlogAttachmentResults(0)),
      ),
    );
    searchSpaces.mockReturnValue(
      CancellablePromise.from(Promise.resolve(createSpaceResponse(0))),
    );
  };

  const waitForRecents = async () => {
    try {
      await getRecentPeople().promise();
    } catch (e) {}

    try {
      await getRecentItems().promise();
    } catch (e) {}

    try {
      await getRecentSpaces().promise();
    } catch (e) {}
  };

  const searchServices = jest.fn();
  const searchLibsAppsMore = jest.fn();
  const searchTeamsAndUsers = jest.fn();

  const searchDetails = {
    data: undefined,
    loading: false,
    error: null,
  };

  const waitForSearch = async (debounceTime: number) => {
    try {
      await searchPeople().promise();
    } catch (e) {}

    try {
      await searchItems().promise();
    } catch (e) {}

    try {
      await searchSpaces().promise();
    } catch (e) {}

    try {
      await searchServices().promise();
    } catch (e) {}

    try {
      await searchLibsAppsMore().promise();
    } catch (e) {}

    try {
      await searchTeamsAndUsers().promise();
    } catch (e) {}

    // Await an extra loop to account for the debounce time, this is required even if the debounce is 0
    await new Promise<void>((resolve) =>
      setTimeout(() => resolve(), debounceTime),
    );
  };

  const mockClients = {
    recentClient: {
      getRecentItems,
      getRecentSpaces,
    },
    searchClient: {
      search: jest.fn(() => ({
        [Scope.ConfluencePageBlogAttachment]: searchItems(),
        [Scope.ConfluenceSpace]: searchSpaces(),
        [Scope.People]: searchPeople(),
      })),
      getRecentPeople,
    },
  } as {
    recentClient: any;
    searchClient: any;
  };

  const onOpen = jest.fn();
  const onClose = jest.fn();
  const setAdditionalAnalyticsContext = jest.fn();
  const onRetry = jest.fn();
  const requestComplete = jest.fn();

  const commonProps = {
    onOpen,
    onClose,
    setAdditionalAnalyticsContext,
    onRetry,
    requestComplete,
    forwardRef: () => {},
    searchSessionId: 'SOME_SEARCH_SESSION_ID',
    debounceTime: 0,
    query: '',
    queryVersion: 0,
    enabledFilters: [],
    allSiteFilters: [],
    servicesQuery: {
      searchServices,
      fetchMore,
      teamNames: {},
      ...searchDetails,
    },
    libsAppsMoreQuery: {
      searchLibsAppsMore,
      fetchMore,
      teamNames: {},
      ...searchDetails,
    },
    teamsAndUsersQuery: { searchTeamsAndUsers, ...searchDetails, data: [] },
    recentlyViewedComponents: { loading: false, components: [] },
    recentlyViewedTeams: { loading: false, teams: [] },
  };

  const sites = createSiteFilters(2, {
    checkedNumber: 2,
  });

  beforeEach(() => {
    jest.clearAllMocks();
    mockSuccessRecent();
    mockSuccessSearch();
  });

  it("doesn't render if there are no results in pre query", () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog {...commonProps} isExpanded {...mockClients} />,
    );

    expect(wrapper.find(SearchError).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        recentlyViewedComponents={{
          loading: false,
          components: [
            {
              id: 'test-comp',
              name: 'microscope',
              type: CompassComponentType.APPLICATION,
              __typename: 'CompassComponent',
            },
          ],
        }}
        isExpanded
        {...mockClients}
      />,
    );

    expect(wrapper.find(SearchError).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it("doesn't render pre query results if defaultSearchState is not enabled", () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog {...commonProps} isExpanded {...mockClients} />,
    );

    expect(wrapper.find(SearchError).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render dialog when no pre query results', async () => {
    mockEmptyRecents();

    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded={false}
        {...mockClients}
      />,
    );

    // Wait for the requests to resolve, swallowing any errors
    await waitForRecents();

    wrapper.update();

    expect(wrapper.find(ConfluenceSearchDialog).exists()).toBe(false);
  });

  it('matches snapshot and renders error when there is an error when searching', async () => {
    mockErrorSearch();

    const commonProps = {
      onOpen,
      onClose,
      setAdditionalAnalyticsContext,
      onRetry,
      requestComplete,
      forwardRef: () => {},
      searchSessionId: 'SOME_SEARCH_SESSION_ID',
      debounceTime: 0,
      query: '',
      queryVersion: 0,
      enabledFilters: [],
      allSiteFilters: [],
      servicesQuery: {
        searchServices,
        data: undefined,
        error: new Error(''),
        loading: false,
        fetchMore,
        teamNames: {},
      },
      libsAppsMoreQuery: {
        searchLibsAppsMore,
        fetchMore,
        ...searchDetails,
        teamNames: {},
        error: new Error(''),
      },
      teamsAndUsersQuery: {
        searchTeamsAndUsers,
        ...searchDetails,
        error: new Error(''),
        data: [],
      },
      recentlyViewedComponents: { loading: false, components: [] },
      recentlyViewedTeams: { loading: false, teams: [] },
    };

    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded
        {...mockClients}
        debounceTime={0}
      />,
    );

    await waitForRecents();

    wrapper.setProps({ query: 'some query' });

    await waitForSearch(0);

    wrapper.update();

    expect(wrapper.find(SearchError).exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot and does not render error when there are no results when searching', async () => {
    mockEmptySearch();

    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded
        {...mockClients}
        debounceTime={0}
      />,
    );

    await waitForRecents();

    wrapper.setProps({ query: 'some query' });

    await waitForSearch(0);

    wrapper.update();

    expect(wrapper.find(SearchError).exists()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('passes the right parameters to the search client when making a query', async () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded
        {...mockClients}
        debounceTime={0}
      />,
    );

    wrapper.setProps({ query: 'some query' });

    await waitForSearch(0);

    expect(searchServices).toBeCalledTimes(2);
    expect(searchServices).toBeCalledWith('some query', null);
  });

  it('passes the right parameters to the search client when making a query with a selected site', async () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded
        {...mockClients}
        debounceTime={0}
        allSiteFilters={sites}
      />,
    );

    wrapper.setProps({ query: 'some query' });

    await waitForSearch(0);

    expect(searchServices).toBeCalledTimes(2);
    expect(searchServices).toBeCalledWith('some query', null);
  });

  it('correctly resets to pre-query when closed', async () => {
    const wrapper: ShallowWrapper<any, State> = shallow(
      <ConfluenceSearchDialog {...commonProps} isExpanded {...mockClients} />,
    );

    // Change query state to non pre-query
    wrapper.setProps({ query: 'some query' });
    wrapper.update();

    await waitForSearch(0);

    // Close the dialog
    wrapper.setProps({
      isExpanded: false,
      query: '',
    });

    wrapper.update();

    // Assert
    expect(wrapper.state('isPreQuery')).toBe(true);
  });

  it('sends analytics when fetching more', () => {
    const wrapper = shallow(
      <ConfluenceSearchDialog
        {...commonProps}
        isExpanded
        {...mockClients}
        createAnalyticsEvent={mockCreateAnalyticsEvent}
        recentlyViewedComponents={{
          loading: false,
          components: [
            {
              id: 'test-comp',
              name: 'microscope',
              type: CompassComponentType.APPLICATION,
              __typename: 'CompassComponent',
            },
          ],
        }}
      />,
    );

    const fetchMore = wrapper.find('_default').prop('fetchMoreFn');
    (fetchMore as any)('SERVICES', '390rtjgawp3yh');

    expect(mockCreateAnalyticsEvent).toBeCalledWith({
      action: 'clicked',
      actionSubject: 'showMoreButton',
    });
    expect(fireUIAnalytics).toBeCalledWith(mockEvent, 'compass-services');

    mockCreateAnalyticsEvent.mockClear();
    (fireUIAnalytics as any).mockClear();

    (fetchMore as any)('LIBS_APPS_MORE', '4nt9g0famw3kyc');

    expect(mockCreateAnalyticsEvent).toBeCalledWith({
      action: 'clicked',
      actionSubject: 'showMoreButton',
    });
    expect(fireUIAnalytics).toBeCalledWith(mockEvent, 'compass-libsApps');
  });

  it('checks analytics on search completion', () => {
    shallow(
      <WrappedConfluenceSearchDialog
        isExpanded={true}
        debounceTime={20}
        query={'testQuery'}
        onRetry={() => {}}
        setAdditionalAnalyticsContext={() => {}}
        queryVersion={1}
      />,
    );

    expect((useSearchComponentsLazyQuery as any).mock.calls.length).toEqual(2);
    // Check analytics event for services search completion
    (useSearchComponentsLazyQuery as any).mock.calls[0][0].onCompleted({
      compass: {
        searchComponents: {
          nodes: ['service1', 'service2', 'service3', 'service4'],
          __typename: 'CompassSearchComponentConnection',
        },
      },
    });
    expect(fireTrackAnalytics).toBeCalledWith({}, 'searchRequest completed', {
      actionSubjectId: 'compass-services',
      resultCount: 4,
    });

    // Check analytics event for libs + apps search completion
    (useSearchComponentsLazyQuery as any).mock.calls[1][0].onCompleted({
      compass: {
        searchComponents: {
          nodes: ['lib', 'lib2', 'app1', 'app2', 'app3'],
          __typename: 'CompassSearchComponentConnection',
        },
      },
    });
    expect(fireTrackAnalytics).toBeCalledWith({}, 'searchRequest completed', {
      actionSubjectId: 'compass-libsApps',
      resultCount: 5,
    });

    // Check analytics event for teams search completion
    expect((useSearchTeams as any).mock.calls.length).toEqual(1);
    (useSearchTeams as any).mock.calls[0][4](['team1', 'team2']);
    expect(fireTrackAnalytics).toBeCalledWith({}, 'searchRequest completed', {
      actionSubjectId: 'compass-teams',
      resultCount: 2,
    });

    // Check analytics event for users search completion
    expect((useSearchUsers as any).mock.calls.length).toEqual(1);
    (useSearchUsers as any).mock.calls[0][0].onCompleted(['user1']);
    expect(fireTrackAnalytics).toBeCalledWith({}, 'searchRequest completed', {
      actionSubjectId: 'compass-users',
      resultCount: 1,
    });
  });
});
