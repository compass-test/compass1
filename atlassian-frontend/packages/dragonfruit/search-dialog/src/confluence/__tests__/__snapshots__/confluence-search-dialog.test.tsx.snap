// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConfluenceSearchDialog /> doesn't render if there are no results in pre query 1`] = `""`;

exports[`<ConfluenceSearchDialog /> doesn't render pre query results if defaultSearchState is not enabled 1`] = `""`;

exports[`<ConfluenceSearchDialog /> matches snapshot 1`] = `
<Component>
  <Component>
    <_default
      fetchMoreFn={[Function]}
      isBelowTakeoverWidth={false}
      isPreQuery={true}
      libsAppsMore={
        Object {
          "cursor": null,
          "isLoading": false,
          "results": Array [],
        }
      }
      people={
        Object {
          "isLoading": true,
          "results": null,
        }
      }
      peopleTeams={
        Object {
          "isLoading": false,
          "results": Array [],
        }
      }
      query=""
      recentlyViewedComponents={
        Object {
          "components": Array [
            Object {
              "__typename": "CompassComponent",
              "id": "test-comp",
              "name": "microscope",
              "type": "APPLICATION",
            },
          ],
          "loading": false,
        }
      }
      recentlyViewedTeams={
        Object {
          "loading": false,
          "teams": Array [],
        }
      }
      services={
        Object {
          "cursor": null,
          "isLoading": false,
          "results": Array [],
        }
      }
      showFilters={false}
      teamNames={Object {}}
    />
  </Component>
</Component>
`;

exports[`<ConfluenceSearchDialog /> matches snapshot and does not render error when there are no results when searching 1`] = `
<Component>
  <Component>
    <_default
      fetchMoreFn={[Function]}
      isBelowTakeoverWidth={false}
      isPreQuery={false}
      libsAppsMore={
        Object {
          "cursor": null,
          "isLoading": false,
          "results": Array [],
        }
      }
      people={
        Object {
          "isLoading": true,
          "results": null,
        }
      }
      peopleTeams={
        Object {
          "isLoading": false,
          "results": Array [],
        }
      }
      query="some query"
      recentlyViewedComponents={
        Object {
          "components": Array [],
          "loading": false,
        }
      }
      recentlyViewedTeams={
        Object {
          "loading": false,
          "teams": Array [],
        }
      }
      services={
        Object {
          "cursor": null,
          "isLoading": false,
          "results": Array [],
        }
      }
      showFilters={true}
      teamNames={Object {}}
    />
  </Component>
</Component>
`;

exports[`<ConfluenceSearchDialog /> matches snapshot and renders error when there is an error when searching 1`] = `
<Component>
  <Component>
    <div
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <InjectIntl(SearchError)
        onRetry={[Function]}
      />
    </div>
  </Component>
</Component>
`;
